Changes necessary to support `netctl login`

dseevr opened the pull request:
These are the changes from https://github.com/contiv/modelgen/pull/24.  Regular commit message follows:

This commit does not break any existing functionality for regular netmaster communication

Changes:

- default http client is replaced by a http.Client in the ContivClient struct
- httpGet()/httpPost()/httpDelete() are moved into ContivClient and use the new internal http.Client
- added SetHttpClient() which allows for overriding the internal http.Client (to disable cert checking, add timeouts, etc.)
- added Login() function which performs a login POST request against auth_proxy
- added SetAuthToken() which adds a custom request header (X-Auth-Token) before requests are sent out
- httpGet()/httpPost()/httpDelete() are modified to set any custom request headers before performing the request

Signed-off-by: Bill Robinson <dseevr@users.noreply.github.com>

yuva29 commented on client/contivModelClient.go:
```diff
@@ -80,7 +136,16 @@ type Link struct {
+		return err
+	}
+
+	if err := addAuthTokenHeader(req); err != nil {
```
Are we enforcing that all the requests should go through proxy? What if the user wants to bypass proxy and make a call to netmaster directly?

dseevr replied:
See line 95.  Non-HTTPS requests return early and don't get the header added (for security reasons) so this doesn't interfere with normal operation

Non-TLS netmasters can still be talked to normally (omit `--netmaster` for the default on `localhost:9999` or specify one manually)

yuva29 replied:
Got it.

rhim reviewed (approved):
LGTM.

jojimt commented on client/contivModelClient.go:
```diff
@@ -64,14 +64,70 @@ package client
+
+type NetctlConfig struct {
+	Token string `json:"token"`
+}
```
Let's separate out the netctl specific items and move them out of the model as discussed in the modelgen PR.

rhim reviewed (commented):
why are there changes for check package in this patch?

dseevr committed:
Changes necessary to support `netctl login`

Changes:

- default http client is replaced by a http.Client in the ContivClient struct
- httpGet()/httpPost()/httpDelete() are moved into ContivClient and use the new internal http.Client
- added SetHttpClient() which allows for overriding the internal http.Client (to disable cert checking, add timeouts, etc.)
- added Login() function which performs a login POST request against auth_proxy
- added SetAuthToken() which adds a custom request header (X-Auth-Token) before requests are sent out
- httpGet()/httpPost()/httpDelete() are modified to set any custom request headers before performing the request
- added `make systemtests` target which runs everything under the systemtests dir

Signed-off-by: Bill Robinson <dseevr@users.noreply.github.com>

dseevr committed:
Godeps changes to bring in contiv's fork of gocheck

Signed-off-by: Bill Robinson <dseevr@users.noreply.github.com>

dseevr committed:
RAML changes from build process

Signed-off-by: Bill Robinson <dseevr@users.noreply.github.com>

dseevr commented:
@rhim that was from bringing in contiv's fork of gocheck.  I broke that out into a separate commit

rhim reviewed (approved):
These changes look fine. I am a bit unclear on tests that we need for auth_proxy - basically test that ensures that no command can be run without token.

Another test would be that of invalid/expired token. 

As long as we have those scenarios covered, I think this is good to go.

Please also open the doc change PR for this change.

dseevr commented:
@rhim there's no problem with netctl sending requests to auth_proxy without a token (e.g., if they haven't logged in yet or deleted the config after logging in) because they will get back a 401 unauthorized.  auth_proxy has tests to cover that scenario.  It also has tests to cover invalid and expired tokens.

Do you want netctl to refuse to send the requests if the target is https and no token (i.e., config) is present?

rhim commented:
No, that’s ok. Shouldn’t overburden client if server side mechanism already exists.

- Himanshu
     .:|:.:|:.

On May 4, 2017, at 5:29 PM, Bill Robinson <notifications@github.com<mailto:notifications@github.com>> wrote:


@rhim<https://github.com/rhim> there's no problem with netctl sending requests to auth_proxy without a token (e.g., if they haven't logged in yet or deleted the config after logging in) because they will get back a 401 unauthorized. auth_proxy has tests to cover that scenario. It also has tests to cover invalid and expired tokens.

Do you want netctl to refuse to send the requests if the target is https and no token (i.e., config) is present?

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub<https://github.com/contiv/contivmodel/pull/64#issuecomment-299344759>, or mute the thread<https://github.com/notifications/unsubscribe-auth/AMiMKYjK_Iyb4MHEI-QDl7IX-YwkVecpks5r2m1ZgaJpZM4M6uy9>.
